projects/vdk-plugins/vdk-jupyter/vdk-jupyterlab-extension/src/components/Login2.tsx

```diff
@@ -0,0 +1,148 @@
+ * Since notebooks can be started using random URL we cannot speicfy fixed callback URL.
+ *
+ * This component provides an interactive method for users to authenticate
+ * through an external OAuth provider.
+ *
+ * Workflow:
+ * 1. An authentication window is opened for the user upon invoking `openAuthWindow`.
+ *    - This makes a request to the 'login' endpoint with the origin specified.
+ *    - The server responds with a redirect URL, which is then opened in a new window.
+ *
+ * 2. After the user completes the authentication on the OAuth provider's site,
+ *    they are redirected back, with an authorization code in the URL.
+ *
+ * 3. Users are instructed to copy and paste this URL into an input field manualy
+ *
+ * 4. This code is then sent to the 'login' endpoint to exchange it for tokens or finalize the authentication process.
+ *
+ * 5. The component's state is updated based on the success or failure of the authentication.
+ *
+ */
+export class ManualOAuth2Flow extends Component<{}, ManualOAuth2FlowState>  {
+    private authWindow: Window | null = null;
+
+    constructor(props: {}) {
+        super(props);
+        this.state = {
+            authStatus: AuthStatus.NONE
+        };
+    }
+
+
+    openAuthWindow = async () => {
+        const redirect_url = await requestAPI<any>(
+            'login',
+            {
+                method: 'GET'
+            }
+        );
+
+        this.authWindow = window.open(redirect_url, 'authWindow', 'width=600,height=400');
+        this.setState({authStatus: AuthStatus.WINDOW_OPENED});
+    }
+
+    closeAuthWindow = () => {
+        if (this.authWindow && !this.authWindow.closed) {
+            this.authWindow.close();
+        }
+    }
+
+    handleURLInput = async (e: { target: { value: any; }; }) => {
+        const redirectURI = e.target.value;
+        const code = new URL(redirectURI).searchParams.get('code');
+        if (code) {
+            try {
+                await requestAPI<any>(
+                    'login?code=' + code,
+                    {
+                        method: 'GET'
+                    }
+                );
+                this.closeAuthWindow();
+                this.setState({ authStatus: AuthStatus.SUCCESS });
+            } catch (error) {
+                this.closeAuthWindow();
+                console.error("Error during authentication:", error);
+                this.setState({ authStatus: AuthStatus.FAILURE });
+            }
+        }
+    }
+
+    render() {
+        return (
+            <div className={LOGIN2_DIALOG_CLASS}>
+                {this.state.authStatus === AuthStatus.NONE &&
+                    <div>
+                        <p>Authenticate process will start. After you clik a new window will be opened.</p>
+                        <ul>
+                            <li>Please <b>wait</b> for the authentication window to redirect you back to 127.0.0.1.</li>
```

Comment:
Why is it specifically 127.0.0.1?

Reply:
In the OAuth2 authorization flow, the redirect URI is where the user is sent after they have authenticated with the authorization server. This URI must be registered with the OAuth2 provider and must match exactly the redirect URI sent in the authorization request.

Using 127.0.0.1 as the redirect URI is a fairly safe option as it always point to your local machine. This address can't be hijacked with DNS changes.